### PR TITLE
Using new parquet in train

### PR DESCRIPTION
**Pull request by cbutsko**

### CODE REVIEW

File: paper_eval.py [hunk 1]
```diff
@@ -89,7 +102,21 @@
...
+
+logger.info("Reading dataset")
+files = sorted(glob(f"{parquet_file}/**/*.parquet"))[:10]
+df_list = []
+for f in tqdm(files):
+    _data = pd.read_parquet(f, engine="fastparquet")
+    _data_pivot = process_parquet(_data)
+    _data_pivot.reset_index(inplace=True)
+    df_list.append(_data_pivot)
+    del _data, _data_pivot
```
Comment (kvantricht) [inline paper_eval.py]: does this truly have impact? In the next iteration this variable anyway gets overwritten?

File: paper_eval.py [hunk 2]
```diff
@@ -104,13 +131,14 @@
...
-model_modes = ["Random Forest", "Regression", "CatBoostClassifier"]
+# model_modes = ["Random Forest", "Regression", "CatBoostClassifier"]
+model_modes = ["CatBoostClassifier"]
 
 # 1. Using the provided split
 val_samples_df = pd.read_csv(data_dir / val_samples_file)
 train_df, test_df = WorldCerealBase.split_df(df, val_sample_ids=val_samples_df.sample_id.tolist())
 full_eval = WorldCerealEval(
-    train_df, test_df, spatial_inference_savedir=model_logging_dir, dekadal=dekadal
+    train_df, test_df, spatial_inference_savedir=model_logging_dir, dekadal=dekadal, augment=False
```
Comment (kvantricht) [inline paper_eval.py]: shouldn't this value be set by the `augment` variable instead of hardcoded to `False`?

File: presto/dataset.py [hunk 3]
```diff
@@ -65,23 +65,73 @@ def target_crop(row_d: Dict) -> int:
...
+    @classmethod
+    def get_timestep_positions(cls, row_d: Dict, augment: bool = False) -> List[int]:
+        available_timesteps = int(row_d["available_timesteps"])
+        valid_position = int(row_d["valid_position"])
+
+        # force moving the center point if it is too close to the edges
+        if (valid_position < cls.NUM_TIMESTEPS // 2) or (
+            valid_position > (available_timesteps - cls.NUM_TIMESTEPS // 2)
+        ):
+            augment = True
```
Comment (kvantricht) [inline presto/dataset.py]: In this case we enter unconsciously in augmentation which might not be what we desire. Can't we put this logic inside `if not augment` and in that case not choose the `valid_position` as the `center_point` but rather the the point that keeps `valid_position` closest as possible to the center? Then it's always deterministic and we don't have to force going through the augmentation part.

File: presto/dataset.py [hunk 4]
```diff
@@ -65,23 +65,73 @@ def target_crop(row_d: Dict) -> int:
...
         row_d = pd.Series.to_dict(row)
 
         latlon = np.array([row_d["lat"], row_d["lon"]], dtype=np.float32)
-        month = datetime.strptime(row_d["start_date"], "%Y-%m-%d").month - 1
+
+        timestep_positions = cls.get_timestep_positions(row_d, augment=augment)
+        # make sure that month for encoding gets shifted according to
+        # the selected timestep positions
+        month = (
+            pd.to_datetime(row_d["start_date"]) + pd.DateOffset(months=timestep_positions[0])
```
Comment (kvantricht) [inline presto/dataset.py]: You could also just add `timestep_positions[0]` and not use the overhead of the `pd.DateOffset` method. I don't think we're resilient in any case to situations where we're working with dekadal data. This is a potential risk if someone (including ourselves) is going over the dekadal track. Can we make it more universal here? Or if needed for now do a check and raise something, so nobody blindly takes this while it won't work as expected.

File: presto/eval.py [hunk 5]
```diff
@@ -278,7 +281,7 @@ def evaluate(
         pretrained_model: Optional[PrestoFineTuningModel] = None,
     ) -> Dict:
 
-        test_ds = self.ds_class(self.test_df, target_function=self.target_function)
+        test_ds = self.ds_class(self.test_df, target_function=self.target_function, augment=False)
```
Comment (kvantricht) [inline presto/eval.py]: you mean no augmentation for inference? This is also the default I guess?

File: presto/dataset.py [hunk 6]
```diff
@@ -275,6 +326,7 @@ def __init__(
             dataframe = dataframe[(~dataframe.end_date.dt.year.isin(years_to_remove))]
         self.target_function = target_function if target_function is not None else self.target_crop
         self._class_weights: Optional[np.ndarray] = None
+        self.augment = augment
```
Comment (kvantricht) [inline presto/dataset.py:351]: should we add logging somewhere that shows that augmentation is enabled when initializing a dataset like that?

File: presto/utils.py [hunk 7]
```diff
@@ -33,6 +34,9 @@
 device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
 DEFAULT_SEED: int = 42
 
+NODATAVALUE: int = 65535
+MIN_EDGE_BUFFER: int = 2  # Min amount of timesteps to include before/after the valid position
```
Comment (kvantricht) [inline presto/utils.py]: we either have to make this a more informative variable name or move it to where we need it (only in dataset.py?) Is this only used in case of augmentation? Then we could also add this to the name. TBD.

File: presto/utils.py [hunk 8]
```diff
@@ -79,6 +83,237 @@ def timestamp_dirname(suffix: Optional[str] = None) -> str:
...
+    Raises
+    ------
+    ValueError
+        error is raised if pivot results in an empty DataFrame
+    """
+
+    # add dummy value + rename stuff for compatibility with existing functions
+    df["OPTICAL-B8A"] = NODATAVALUE
+
+    # TODO: this needs to go away once the transition to new data is complete
```
Comment (kvantricht) [inline presto/utils.py:127]: yes, and can we immediately go to presto naming convention then as well?

**Review by kvantricht — COMMENTED** [review]
Good work! Added already some comments. The `process_parquet` is something that's for me hard to read just as code. Let's start by testing it thoroughly.

File: presto/dataset.py (hunk 6 above)
Comment (cbutsko) [reply presto/dataset.py:351]: good point. addressed that here ae27e25db6a172bfac1c1fe0fa3d49861d3780d4

File: presto/eval.py (hunk 5 above)
Comment (cbutsko) [reply presto/eval.py]: right. removed this parameter 8d7e4c11876356a360f693805c6c0444c3fcd09d

File: presto/dataset.py (hunk 3 above)
Comment (cbutsko) [reply presto/dataset.py]: good point. tried to rewrite it here b2f1aa383e60eaf3bb6f308083d75a2a5c71e9eb

File: paper_eval.py (hunk 2 above)
Comment (cbutsko) [reply paper_eval.py]: 🤦‍♀️ fixed ff255099e5ead303211fac3442cb6fef3f2e49c5

File: paper_eval.py (hunk 1 above)
Comment (cbutsko) [reply paper_eval.py]: fixed ff255099e5ead303211fac3442cb6fef3f2e49c5

File: presto/utils.py (hunk 7 above)
Comment (cbutsko) [reply presto/utils.py]: moved these two to dataops.py, since they are used both in dataset.py and utils.py. looks like it is the best place for now, since circular imports are avoided in this case, and semantically dataops.py contain similar parameters e338c0915f78e59bfa6074f64627b9d933b0752d

File: presto/dataset.py (hunk 4 above)
Comment (cbutsko) [reply presto/dataset.py]: good point. tried to account for that here ad05b3de8acbd5379e776fdb92969e484bd1da3c
we will also need to think about renaming this `month` thing (since it can be something else too), and also making sure that other relative timestep positions (`valid_position` and `timestep_ind`) are computed not as month, but in a more generic fashion

Comment (kvantricht) [reply presto/dataset.py]: You mean this has to be tackled in `process_parquet`?

Comment (kvantricht) [reply presto/dataset.py]: I am not sure I understand what `initial_start_date_position` actually means and why it cannot just be the month inferred from `start_date`.

Comment (cbutsko) [reply presto/dataset.py]: well, maybe I'm overthinking it...

here's an example:
1. we are working on a monthly basis, with start_date in October (hence, `month = 10`). 
and we want to shift it 4 timesteps forward. we can just make `(10 + 4)` and add the modulo part not to get bad month value `% cls.NUM_TIMESTEPS`

2. we are working on a dekadal basis. start date is the same. I assume that our timestep indices are not in month chunks, but in 10-day intervals. so, we have observations for ts0, ts1, ..., ts45 (for example). our `NUM_TIMESTEPS` variable should be set to 36 instead of 12 (like Giorgia did). the `valid_date` should also translate into 10-day chunk instead of month. so that we can select 36 timesteps around the `valid_position`. 
now, we have our `timestep_positions` returned in dekadal steps. so when we add 4 dekadal steps to a month value, it doesn't make sense. we need to add apples to apples to get the date that accounts for the shift.
and now we can take it's month and pass it to presto. I realize now that this particular step is missing in my implementation.

am I missing something?

Comment (kvantricht) [reply presto/dataset.py]: Following most of it. But still, `start_date` is a real date, from which we can infer what would normally be the start month being fed to Presto. Why do we need to translate it to a position? I agree that what we have to add to it due to the shift should account for the time "resolution". `valid_position` should be the translation of `valid_date` (which is irrespective of time resolution) to the position in the timesteps (which is depending on the time resolution). I might be overthinking it just as much.

Comment (cbutsko) [reply presto/dataset.py]: okay, maybe something like this can work:
```
step_converted_to_month = np.ceil(timestep_positions[0] * (365 // NUM_TIMESTEPS) / 30)
month = (pd.to_datetime(start_date).month + step_converted_to_month) % 12 - 1
```
it can be a little imprecise in some cases, not more than 1 month error.
but it's just a one-liner, so probably we can sacrifice this little bit of precision

Comment (kvantricht) [reply presto/dataset.py]: yeah let's make a note that Giorgia can test this out properly when doing dekadal runs.

**Comment by kvantricht** [comment]
@cbutsko is this branch already part of croptype? You were training on new parquet format already? But maybe only locally?

**Comment by cbutsko** [comment]
@kvantricht this branch is not yet the part of croptype. My plan is to sync it now with main, and then merge it directly into croptype, as you suggested.

File: paper_eval.py
Comment (kvantricht) [inline paper_eval.py:1]: @cbutsko in `croptype` branch this file is removed. What does that mean for the changes here?

File: presto/dataset.py [hunk 9]
```diff
@@ -32,7 +34,7 @@
 
 
 class WorldCerealBase(Dataset):
-    _NODATAVALUE = 65535
+    # _NODATAVALUE = 65535
```
Comment (kvantricht) [inline presto/dataset.py:37]: @cbutsko why is this commented? And are changes below like the addition of `get_timestep_positions` not part of the other branch?

File: presto/dataset.py [hunk 10]
```diff
@@ -382,7 +460,24 @@ def __getitem__(self, idx):
 
 
 class WorldCerealInferenceDataset(Dataset):
-    _NODATAVALUE = 65535
+    # _NODATAVALUE = 65535
+    Y = "worldcereal_cropland"
+    BAND_MAPPING = {
```
Comment (kvantricht) [inline presto/dataset.py:466]: these changes will all conflict with croptype branch I think?

File: paper_eval.py
Comment (cbutsko) [reply paper_eval.py:1]: I merged the functionality of `paper_eval` into `train.py`. I think in this branch it doesn't really matter. I just wanted to align this branch with main and then merge into croptype, resolving these conflict there. also, now after discussing ss training with Giorgia, it seems more feasible to create two separate files with more clear functions: something like `train_self_supervised.py` and `train_finetuned.py`

Comment (kvantricht) [reply paper_eval.py:1]: That sounds like a good plan!

File: presto/dataset.py [hunk 11]
```diff
@@ -63,26 +65,97 @@ def target_crop(row_d: Dict) -> int:
         # by default, we predict crop vs non crop
         return int(row_d["LANDCOVER_LABEL"] == 11)
 
+    @classmethod
+    def get_timestep_positions(cls, row_d: Dict, augment: bool = False) -> List[int]:
+        available_timesteps = int(row_d["available_timesteps"])
+        valid_position = int(row_d["valid_position"])
```
Comment (gabrieltseng) [inline presto/dataset.py:71]: Could you add a comment describing what `valid_position` represents?

File: presto/dataset.py [hunk 12]
```diff
@@ -63,26 +65,97 @@ def target_crop(row_d: Dict) -> int:
...
+
+        if not augment:
```
Comment (gabrieltseng) [inline presto/dataset.py:73]: Could you also make a note that augment here means temporal jittering? Am I right in understanding `MIN_EDGE_BUFFER` is just used to determine how much temporal jittering is allowed? In this case maybe it makes sense as a class attribute (or even as a default argument to this function)

File: presto/eval.py [hunk 13]
```diff
@@ -306,7 +311,7 @@ def evaluate(
         test_preds_np = test_preds_np >= self.threshold
         prefix = f"{self.name}_{finetuned_model.__class__.__name__}"
 
-        catboost_preds = test_ds.df.worldcereal_prediction
+        catboost_preds = test_ds.df.worldcereal_prediction == 11
```
Comment (gabrieltseng) [inline presto/eval.py:314]: What is the motivation for this change?

File: presto/dataset.py (hunk 11 above)
Comment (gabrieltseng) [reply presto/dataset.py:71]: I see its created in `utils.py` but its still not obvious to me what it represents - could you add a comment there instead?

File: catboost_info/catboost_training.json [hunk 14]
```diff
@@ -0,0 +1,1557 @@
+{
```
Comment (gabrieltseng) [inline catboost_info/catboost_training.json:1]: I assume this was an accidental commit?

**Comment by gabrieltseng** [comment]
Nice job @cbutsko - threw in my own comments too. +1 to @kvantricht 's point that the process_parquet is doing a lot, but the docstring is really helpful.

**Comment by kvantricht** [comment]
CHanges are already in croptype branch.
